Automattic/jetpack
# Donations: Save fallback for non-frontend contexts

Fixes https://github.com/Automattic/jetpack/issues/17358
Fixes 241-gh-dotcom-manage
Fixes https://github.com/Automattic/jetpack/issues/17359

#### Changes proposed in this Pull Request:
This PR continues the work initiated in #17256 and adapts the Donations block so the `save` function is used as a proper fallback for non-frontend contexts while `render_callback` is used for augmenting the content dynamically for frontend contexts.

The block saves now a fallback output into the post content. That fallback includes a textual call-to-action linking to the post so folks can donate from the website.
The block also sets a render callback that augments the UI with working donate buttons.

#### Jetpack product discussion
N/A.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
_Web_
- [Spin up a JN site running `master`](https://jurassic.ninja/create/?jetpack-beta).
- Set up Jetpack and start for free.
- You need to sandbox the WP.com store now:
  - Edit your `/etc/hosts` file and point `public-api.wordpress.com` and `subscribe.wordpress.com` to your WP.com sandbox IP.
  - Go to WP Admin > Settings > Jetpack Constants and enter your WP.com sandbox address in `JETPACK__SANDBOX_DOMAIN`.
  - Add the code below to `/wp-content/mu-plugins/0-sandbox.php` in your WP.com sandbox:
```
define( 'USE_STORE_SANDBOX', true );
define( 'USE_BILLING_SANDBOX', true );
define( 'BILLINGDADDY_SANDBOX', rtrim( 'https://' . `hostname` ) );
```
- Purchase a Jetpack Security plan.
- Go to WP Admin > Posts > Add new.
- Insert a Donations block.
- Connect to Stripe.
- Publish the post.
- Go to WP Admin > Jetpack > Jetpack Beta.
- Switch to the branch of this PR (`update/donations-fallback`).
- Visit the frontend of the post you published before.
- Make sure there are no regressions in the Donations block.
- Go to WP Admin > Posts.
- Edit the post you published before.
- Make sure there are no regressions in the Donations block.
- Open the "More" menu and activate the "Code editor" view.
- Make sure the HTML markup is the fallback output (the one with a link to the post).
<img width="400" alt="Screen Shot 2020-10-20 at 16 28 37" src="https://user-images.githubusercontent.com/1233880/96600585-61133880-12f1-11eb-9e5d-76daeebd50bf.png">

- Switch back to the "Visual editor" view.
- Click on the "Update" button.
- Visit the frontend of the post.
- Make sure the augmented UI is displayed (the one with a working donate button).

_API_
- [Go to the WP.com API console](https://developer.wordpress.com/docs/api/console/).
- Make a `GET` request to `/sites/:site/posts/:post-id` (where `:post-id` is the ID of a post that contains a Donations block).
- Make sure the post content uses the fallback.
<img width="400" alt="Screen Shot 2020-10-20 at 16 42 07" src="https://user-images.githubusercontent.com/1233880/96602402-45109680-12f3-11eb-99c1-934804553eed.png">

_Email, notifications, and reader (only testable in WP.com)_
- Apply D51410-code to your WP.com sandbox.
- Sandbox a WP.com site under a paid plan.
- Go to https://wordpress.com/following/manage.
- Make sure to follow the site you sandboxed above.
- Enable the "Notify me of new posts" option.
<img width="200" alt="Screen Shot 2020-09-28 at 16 25 18" src="https://user-images.githubusercontent.com/1233880/94444997-3b3dbc80-01a7-11eb-850f-47e87d259098.png">

- Go to https://wordpress.com/post.
- Select the site you sandboxed above.
- Publish a post that contains a "Donations" block.
- Grab the IDs of the post and the site where the post has been published.
- In a WP.com sandbox run `php ./bin/subscriptions/send.php <BLOG_ID> post <POST_ID> <YOUR_EMAIL>`.
- Check the email received and make sure it contains the fallback link.
<img width="400" alt="Screen Shot 2020-10-20 at 16 57 26" src="https://user-images.githubusercontent.com/1233880/96604388-59ee2980-12f5-11eb-8096-9d67bf23ae70.png">

- Open the masterbar notifications.
- Make sure it includes the fallback link.
<img width="350" alt="Screen Shot 2020-10-20 at 16 58 01" src="https://user-images.githubusercontent.com/1233880/96604464-6ffbea00-12f5-11eb-9963-1aef6d8ff73f.png">

- Go to Reader.
- Open the post you just published above.
- Make sure it includes the fallback link.
<img width="400" alt="Screen Shot 2020-10-20 at 16 58 29" src="https://user-images.githubusercontent.com/1233880/96604516-8013c980-12f5-11eb-9e4a-b074a31a0f13.png">


#### Proposed changelog entry for your changes:
Donations: Save fallback for non-frontend contexts
